lib/CL/pocl_util.c

```diff
@@ -1046,6 +1047,113 @@ pocl_create_command (_cl_command_node **cmd, cl_command_queue command_queue,
+      for (unsigned i = 0; i < command_buffer->num_queues; ++i)
+        {
+          if (q == command_buffer->queues[i])
+            queue_in_buffer = 1;
+        }
+      POCL_RETURN_ERROR_COND ((!queue_in_buffer), CL_INVALID_COMMAND_QUEUE);
+    }
+  else
+    q = command_buffer->queues[0];
+
+  *command_queue = q;
+  return CL_SUCCESS;
+}
+
+cl_int
+pocl_create_recorded_command (_cl_recorded_command **cmd,
+                              cl_command_buffer_khr command_buffer,
+                              cl_command_queue command_queue,
+                              cl_command_type command_type,
+                              cl_uint num_sync_points_in_wait_list,
+                              const cl_sync_point_khr *sync_point_wait_list,
+                              size_t num_buffers, cl_mem *buffers,
+                              char *readonly_flags)
+{
+  cl_int errcode = pocl_check_syncpoint_wait_list (
+      command_buffer, num_sync_points_in_wait_list, sync_point_wait_list);
+  if (errcode != CL_SUCCESS)
+    return errcode;
+
+  *cmd = (_cl_recorded_command *)calloc (1, sizeof (_cl_recorded_command));
+  _cl_recorded_command *c = *cmd;
+  POCL_RETURN_ERROR_COND ((c == NULL), CL_OUT_OF_HOST_MEMORY);
+  c->type = command_type;
+
+  c->num_sync_points_in_wait_list = num_sync_points_in_wait_list;
+  if (num_sync_points_in_wait_list > 0)
+    {
+      c->sync_point_wait_list = (cl_sync_point_khr *)malloc (
+          num_sync_points_in_wait_list * sizeof (cl_sync_point_khr));
+      POCL_RETURN_ERROR_COND ((c->sync_point_wait_list == NULL
+                               && num_sync_points_in_wait_list != 0),
+                              CL_OUT_OF_HOST_MEMORY);
+      memcpy (c->sync_point_wait_list, sync_point_wait_list,
+              num_sync_points_in_wait_list * sizeof (cl_sync_point_khr));
+    }
+
+  c->memobj_count = num_buffers;
+  if (num_buffers > 0)
+    {
+      c->memobj_list = (cl_mem *)malloc (sizeof (cl_mem) * num_buffers);
+      POCL_RETURN_ERROR_COND ((c->memobj_list == NULL), CL_OUT_OF_HOST_MEMORY);
+      memcpy (c->memobj_list, buffers, sizeof (cl_mem) * num_buffers);
+
+      c->readonly_flag_list = (char *)malloc (sizeof (char) * num_buffers);
+      POCL_RETURN_ERROR_COND ((c->readonly_flag_list == NULL),
+                              CL_OUT_OF_HOST_MEMORY);
+      memcpy (c->readonly_flag_list, readonly_flags,
+              sizeof (char) * num_buffers);
+    }
+
+  return CL_SUCCESS;
+}
+
+cl_int
+pocl_command_record (cl_command_buffer_khr command_buffer,
+                     _cl_recorded_command *cmd, cl_sync_point_khr *sync_point)
+{
+  POCL_LOCK (command_buffer->mutex);
+  if (command_buffer->state != CL_COMMAND_BUFFER_STATE_RECORDING_KHR)
+    {
+      POCL_UNLOCK (command_buffer->mutex);
+      return CL_INVALID_OPERATION;
+    }
+  LL_APPEND (command_buffer->cmds, cmd);
+  if (sync_point != NULL)
+    *sync_point = command_buffer->num_syncpoints;
+  command_buffer->num_syncpoints++;
```

Comment:
Likely not realistic, but 32b value could theoretically overflow. Is there a limit for number of syncpoints per cbuf somewhere?

Reply:
cl_syncpoint is defined as cl_uint which is 32b so it's only possible to have 2^32 unique syncpoints in a command buffer to begin with. I guess the counter could use cl_syncpoint as its type just in case the definition gets changed in the future.